common/mqtt_tests/src/test_mqtt_client.rs

```diff
@@ -0,0 +1,170 @@
+        .await
+        .ok()?;
+    match tokio::time::timeout(timeout, con.next_message()).await {
+        // One collapse both timeout and error to None
+        Err(_) | Ok(Err(_)) => None,
+        Ok(Ok(x)) => Some(x),
+    }
+}
+
+pub struct TestCon {
+    client: AsyncClient,
+    eventloop: EventLoop,
+}
+
+impl TestCon {
+    pub fn new(mqtt_port: u16) -> TestCon {
+        let id: String = std::iter::repeat_with(fastrand::alphanumeric)
+            .take(10)
+            .collect();
+        let mut options = MqttOptions::new(id, "localhost", mqtt_port);
+        options.set_clean_session(true);
+
+        let (client, eventloop) = AsyncClient::new(options, 10);
+        TestCon { client, eventloop }
+    }
+
+    pub async fn subscribe(&mut self, topic: &str, qos: QoS) -> Result<(), anyhow::Error> {
+        self.client.subscribe(topic, qos).await?;
+
+        loop {
+            match self.eventloop.poll().await {
+                Ok(Event::Incoming(Packet::SubAck(_))) => {
+                    return Ok(());
+                }
+                Err(err) => {
+                    return Err(err)?;
+                }
+                _ => {}
+            }
+        }
+    }
+
+    pub async fn publish(
+        &mut self,
+        topic: &str,
+        qos: QoS,
+        payload: &str,
+    ) -> Result<(), anyhow::Error> {
+        self.client.publish(topic, qos, false, payload).await?;
+
+        loop {
+            match self.eventloop.poll().await {
+                Ok(Event::Incoming(Packet::PubAck(_))) => {
+                    return Ok(());
+                }
+                Err(err) => {
+                    return Err(err)?;
+                }
+                _ => {}
+            }
+        }
+    }
+
+    pub async fn forward_received_messages(&mut self, sender: UnboundedSender<String>) {
+        loop {
+            match self.eventloop.poll().await {
+                Ok(Event::Incoming(Packet::Publish(response))) => {
+                    let msg = std::str::from_utf8(&response.payload)
+                        .unwrap_or("Error: non-utf8-payload")
+                        .to_string();
+                    if let Err(_) = sender.send(msg) {
+                        break;
+                    }
+                }
+                Err(err) => {
+                    let msg = format!("Error: {:?}", err).to_string();
+                    let _ = sender.send(msg);
+                    break;
+                }
+                _ => {}
+            }
+        }
+        let _ = self.client.disconnect().await;
+    }
+
+    pub async fn next_message(&mut self) -> Result<String, anyhow::Error> {
+        loop {
+            match self.eventloop.poll().await {
+                Ok(Event::Incoming(Packet::Publish(packet))) => {
+                    let msg = std::str::from_utf8(&packet.payload)
+                        .unwrap_or("Error: non-utf8-payload")
+                        .to_string();
```

Comment:
As this is preventing non-utf8 payloads it may limit usability of the broker, I suppose we don't want to discard binary payloads which are allowed in MQTT.